client/src/main/java/io/pravega/client/auth/CredentialsExtractor.java

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.auth;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.pravega.client.ClientConfig;
+import io.pravega.client.stream.impl.Credentials;
+import java.util.Map;
+import java.util.Properties;
+import java.util.ServiceLoader;
+import java.util.stream.Collectors;
+
+/**
+ * Helper class to extract credentials.
+ */
+public class CredentialsExtractor {
+    private static final String AUTH_PROPS_START = "pravega.client.auth.";
+    private static final String AUTH_METHOD = AUTH_PROPS_START + "method";
+    private static final String AUTH_METHOD_LOAD_DYNAMIC = AUTH_PROPS_START + "loadDynamic";
+
+    private static final String AUTH_PROPS_START_ENV = "pravega_client_auth_";
+    private final Properties properties;
+    private final Map<String, String> env;
+
+    public CredentialsExtractor() {
+        this(System.getProperties(), System.getenv());
+    }
+
+    @VisibleForTesting
+    CredentialsExtractor(Properties properties, Map<String, String> env) {
+        this.properties = properties;
+        this.env = env;
+    }
+
+    /**
+     * Function to extract the credentials object in the given client config.
+     * Here is the order of preference in descending order:
+     * 1. User provides a credential object. This overrides any other settings.
+     * 2. System properties: System properties are defined in the format: "pravega.client.auth.*"
+     * 3. Environment variables. Environment variables are defined under the format "PRAVEGA_CLIENT_AUTH_*"
+     * 4. In case of option 2 and 3, the caller can decide whether the class needs to be loaded dynamically by
+     *     setting property `pravega.client.auth.loadDynamic` to true.
+     *
+     * @param config the config containing the credentials object.
+     * @return Returns a new copy of credentials object. This object is created as per the steps above.
+     */
+    public ClientConfig extract(ClientConfig config) {
```

Comment:
If any new fields are added, there are more places where the builder is used. All the builder code should update for the new field.

Reply:
New fields can be added with default values. This would override them. 
I think we should fix it by having a constructor for the builder that takes an existing instance. That way the update will all be in one class if we add another field.

Reply:
Fixed by adding a builder constructor.

Reply:
I don't see any change